server/mediaserver.go

```diff
@@ -764,29 +764,31 @@ func (s *LivepeerServer) HandlePush(w http.ResponseWriter, r *http.Request) {
 
 		cxn, err = s.registerConnection(st)
 		if err != nil {
-			httpErr := fmt.Sprintf("http push error url=%s err=%v", r.URL, err)
-			glog.Error(httpErr)
-			http.Error(w, httpErr, http.StatusInternalServerError)
-			return
-		}
-
-		// Start a watchdog to remove session after a period of inactivity
-		ticker := time.NewTicker(httpPushTimeout)
-		go func(s *LivepeerServer, intmid, extmid core.ManifestID) {
-			defer ticker.Stop()
-			for range ticker.C {
-				var lastUsed time.Time
-				s.connectionLock.RLock()
-				if cxn, exists := s.rtmpConnections[intmid]; exists {
-					lastUsed = cxn.lastUsed
-				}
-				s.connectionLock.RUnlock()
-				if time.Since(lastUsed) > httpPushTimeout {
-					_ = removeRTMPStream(s, extmid)
-					return
-				}
+			if err != errAlreadyExists {
+				httpErr := fmt.Sprintf("http push error url=%s err=%v", r.URL, err)
+				glog.Error(httpErr)
+				http.Error(w, httpErr, http.StatusInternalServerError)
+				return
 			}
-		}(s, cxn.mid, mid)
+		} else {
+			// Start a watchdog to remove session after a period of inactivity
+			ticker := time.NewTicker(httpPushTimeout)
+			go func(s *LivepeerServer, intmid, extmid core.ManifestID) {
```

Comment:
If an external manifestID foo is provided with the same internal manifestID bar as another external manifestID baz that was previously provided then a watchdog will call `removeRTMPStream()` with baz, but not foo. As a result, `internalManifests` will contain an entry for foo due to L795, but that entry will never be cleared since the [map deletion](https://github.com/livepeer/go-livepeer/blob/7f9534c365123f455c7fd3bfb9ebcbf5a99cdd43/server/mediaserver.go#L577) in `removeRTMPStream()` will never be called for foo.

Reply:
Ah good catch! I also now notice that I'm not cleaning up the `BasicRTMPVideoStream` created for "foo" here as well. Will fix thanks.

Reply:
Oh wait my bad. I think that case won't happen within this PR's context. Different external streamID but same internal manifestID is a separate case altogether, that I've moved to separate issue #1673 and will handle through a separate PR.

Within this PR a segment with a different externalID but same internal manifestID should never even reach L795. The case will get returned with a 500 during `createRTMPStreamIDHandler` itself (for now)

> I'm not cleaning up the BasicRTMPVideoStream created for "foo" here as well. Will fix thanks.
But this still applies for the RTMP struct created for the 'late' segment in the two simultaneous ones arriving for this case. S still a good catch :smile:

Reply:
Also potentially new playlist and new session manager could be created:
```go
	playlist := core.NewBasicPlaylistManager(mid, storage, recordStorage)
```
```go
		sessManager: NewSessionManager(s.LivepeerNode, params, NewMinLSSelector(stakeRdr, 1.0)),
```
both should be cleaned up like in `removeRTMPStream`:
```go
	cxn.stream.Close()
	cxn.sessManager.cleanup()
	cxn.pl.Cleanup()
```

Reply:
IMO it might be easier to read if the function that creates the resource is also responsible for cleaning it up when an error is triggered.

So, in `HandlePush()`

```golang

st := stream.NewBasicRTMPVideoStream(appData)
...

if err != nil {
    st.Close()
    if err != errAlreadyExists {
        ...
    }
}
```

and then in `registerConnection()` we can cleanup the SessionManager and the PlaylistManager if we hit an error condition.